CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI

veronikaslc opened the pull request:
A draft PR untill https://github.com/data-team-uhn/cards/pull/1110 is merged.

Commit:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI

marta- commented on proms-resources/clinical-data/src/main/resources/SLING-INF/content/Homepage/Proms/SurveyInstructions.xml:
```diff
@@ -0,0 +1,52 @@
+    <name>SurveyInstructions</name>
+    <property>
+        <name>eventLabel</name>
+        <value>Upcoming appointment.</value>
```
```suggestion
        <value>Upcoming appointment</value>
```

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/QuestionnaireSet.jsx:
```diff
@@ -482,9 +496,9 @@ function QuestionnaireSet(props) {
+    return (time || location || provider) && surveyInstructions.eventLabel ?
       <Alert severity="info">
-        <AlertTitle>Upcoming appointment</AlertTitle>
+        <AlertTitle>{surveyInstructions.eventLabel}</AlertTitle>
```
I would put the check whether `eventLabel` exists here instead of two lines above.

veronikaslc replied:
You mean just not show the `AlertTitle` if `eventLabel` does not exist?

marta- replied:
Yes. The box still makes sense without the label

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/SurveyInstConfiguration.jsx:
```diff
@@ -0,0 +1,165 @@
+                multiline
+                minRows={4}
+                InputLabelProps={{ shrink: true }}
+                variant="standard"
```
I find `multiline` text fields with `variant="standard"` very hard to look at. Let's make them all outlined.

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/SurveyInstConfiguration.jsx:
```diff
@@ -0,0 +1,165 @@
+            <ListItem key="noSurveysMessage">
+              <TextField
+                InputLabelProps={{ shrink: true }}
+                variant="standard"
```
The current `surveyIntro` and `noSurveysMessage` texts may be short, but we should accommodate potentially longer messages. Let's make these 2 text fields multiline & outlined.

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/SurveyInstConfiguration.jsx:
```diff
@@ -0,0 +1,165 @@
+
+  let buildConfigData = (formData) => {
+    for (let key of Object.keys(surveyInstructions)) {
+      !key.startsWith("jcr:") && formData.append(key, surveyInstructions[key] || defaults[key] || "");
```
This is not the right place to use the defaults. We shouldn't be saving something that the user hasn't entered.

The defaults should be used in `QuestionnaireSet` to make up for missing keys.

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/SurveyInstConfiguration.jsx:
```diff
@@ -0,0 +1,165 @@
+                onChange={(event) => { setSurveyInstructions({...surveyInstructions, interpretationInstructions: event.target.value}); }}
+                fullWidth
+              />
+            </ListItem>
```
There's a lot of copy/paste between these 6 list items. Let's see if we can generate them instead. That means that the `eventLabel` text field may end up to be multiline, which I think is ok. [camelCaseToWords](https://github.com/data-team-uhn/cards/blob/f379774729e02110dfe80fdd69cb709467d71a0f/modules/data-entry/src/main/frontend/src/questionnaireEditor/LabeledField.jsx#L34-L36) can be used to generate the labels.

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/SurveyInstConfiguration.jsx:
```diff
@@ -0,0 +1,165 @@
+        onConfigSaved={() => setHasChanges(false)}
+        >
+          <List>
+            <ListItem key="h5">
+              <Typography variant="h5">Start screen</Typography>
+            </ListItem>
```
I like the Start screen / Summary screen separation 👍

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/SurveyInstConfiguration.jsx:
```diff
@@ -0,0 +1,165 @@
+        onConfigSaved={() => setHasChanges(false)}
+        >
+          <List>
+            <ListItem key="h5">
```
```suggestion
            <ListItem key="startScreen">
```

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/SurveyInstConfiguration.jsx:
```diff
@@ -0,0 +1,165 @@
+                fullWidth
+              />
+            </ListItem>
+            <ListItem key="h52" >
```
```suggestion
            <ListItem key="summaryScreen" >
```

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/QuestionnaireSet.jsx:
```diff
@@ -214,13 +216,25 @@ function QuestionnaireSet(props) {
+    fetch('/Proms/SurveyInstructions.json')
+      .then((response) => response.ok ? response.json() : Promise.reject(response))
+      .then((json) => {
+        setSurveyInstructions(json);
```
This is where we should fill in the gaps with the default instructions.
```suggestion
        setSurveyInstructions(Object.assign(defaultInstructions, json));
```

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/QuestionnaireSet.jsx:
```diff
@@ -533,15 +547,15 @@ function QuestionnaireSet(props) {
-    <Typography color="textSecondary" variant="subtitle1" key="welcome-message">
-        You have no pending surveys to fill out for your next appointment.
-    </Typography>
+    surveyInstructions.noSurveysMessage ? <Typography color="textSecondary" variant="subtitle1" key="welcome-message">
+        {surveyInstructions.noSurveysMessage}
+    </Typography> : null
```
Wondering if there's some creative way we can use a function to generate these instead of repeating  the conditional logic.

Would something like this work? (I haven't tested it)

```
let displayInstructions = (fieldName, Component, props) => (
   surveyInstructions[fieldName] ? <Component {...props}>{surveyInstructions[fieldName]}</Component> : null
);
```
or even more flexible:

```
let displayInstructions = (fieldName, Component, props) => (
   surveyInstructions[fieldName] ?
     Component ?
       <Component {...props}>{surveyInstructions[fieldName]}</Component>
       : surveyInstructions[fieldName]
     : null
);
```

marta- replied:
This hasn't been addressed yet.

marta- replied:
Done in 53cd15b.

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/SurveyInstConfiguration.jsx:
```diff
@@ -0,0 +1,165 @@
+
+  const defaults = {
+    noSurveysMessage: "You have no pending surveys to fill out"
+  };
```
It would be nice to use the defaults as placeholders for the text fields. They can be defined & exported in this file, displayed as placeholders when applicable, and imported by QuestionnaireSet to be used in case there are fields not defined by the user.

```suggestion
 export const defaultInstructions = {
    noSurveysMessage: "You have no pending surveys to fill out"
  };
```

marta- commented on proms-resources/clinical-data/src/main/resources/SLING-INF/content/Homepage/Proms/SurveyInstructions.xml:
```diff
@@ -0,0 +1,52 @@
+    </property>
+    <property>
+        <name>disclaimer</name>
+        <value>Please note that your responses may not be reviewed by your care team until the day of your next appointment. If your symptoms areworsening while waiting for your next appointment, please proceed to your nearest Emergency Department today, or call 911.</value>
```
```suggestion
        <value>Please note that your responses may not be reviewed by your care team until the day of your next appointment. If your symptoms are worsening while waiting for your next appointment, please proceed to your nearest Emergency Department today, or call 911.</value>
```

veronikaslc committed:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI

Label text editing

Co-authored-by: Marta Girdea <marta@phenotips.org>

veronikaslc committed:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI

Better key for the sub-sections separator

Co-authored-by: Marta Girdea <marta@phenotips.org>

veronikaslc committed:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI

Better label for the sub-sections separator

Co-authored-by: Marta Girdea <marta@phenotips.org>

veronikaslc committed:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI

Better place for filling in the gaps with the default instructions

Co-authored-by: Marta Girdea <marta@phenotips.org>

veronikaslc committed:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI

Fixed words gap

Co-authored-by: Marta Girdea <marta@phenotips.org>

Commit:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI
Multiline and outlined view for the TextField

veronikaslc committed:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI

Make defaultsexportable to be reused in other components like QuestionnaireSet

Co-authored-by: Marta Girdea <marta@phenotips.org>

Commit:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI
Right use of default instructions

Commit:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI
Code optimisation; Better empty data alert showing control

veronikaslc commented:
Rebased on the latest CARDS-1836

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/QuestionnaireSet.jsx:
```diff
@@ -214,13 +217,25 @@ function QuestionnaireSet(props) {
+        setError(`Loading the Patient Portal Survey Instructions failed with error code ${response.status}: ${response.statusText}`);
+      });
+  }, []);
+
```
The code is easier to read if the effects are listed in the expected order of their first occurrence. This effect should be first, because it's triggered when the component is mounted.

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/QuestionnaireSet.jsx:
```diff
@@ -482,9 +497,9 @@ function QuestionnaireSet(props) {
     let provider = getVisitInformation("provider");
     provider = provider && provider.length > 1 ? provider.join(", ") : provider;
-    return (time || location || provider) ?
+    return (time || location || provider) &&
```
This accidental/leftover change of `?` into `&&` will cause a syntax error.

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/SurveyInstConfiguration.jsx:
```diff
@@ -0,0 +1,104 @@
+          <List>
+            { Object.keys(labels).map(category => { return (<>
+              <ListItem key={category}>
+                <Typography variant="h5" className={category == "summaryScreen" ? classes.header : ""}>{camelCaseToWords(category)}</Typography>
```
Both are `startScreen` and `summaryScreen` are technically the same kind of headers. You can add the styling to only one of them by putting a classname on the root `Lis`t and selecting only the inner heading with something line `& > > .MuiListItem-root > .MuiTypography-root:not(:first-child)` (actual mui classnames to be double-checked).

Commit:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI
Reorder effects so they are listed on the order they act

Commit:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI
Fixed accidental lefover change

Commit:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI
Better styling code

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/SurveyInstConfiguration.jsx:
```diff
@@ -0,0 +1,106 @@
+	                name={key}
+	                type="text"
+	                label={camelCaseToWords(key)}
+	                value={surveyInstructions ? surveyInstructions[key]|| "" : DEFAULT_INSTRUCTIONS[key] || ""}
```
One of my comments appears to have been lost during rebase, I probably made it directly on the commit directly.

`DEFAULT_INSTRUCTIONS[key]` should be a placeholder. We shouldn't save a value that the user didn't enter.

marta- replied:
Done in 9327acc.

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/SurveyInstConfiguration.jsx:
```diff
@@ -0,0 +1,106 @@
+        buildConfigData={buildConfigData}
+        onConfigSaved={() => setHasChanges(false)}
+        >
+          <List className={classes.header}>
```
I guess my previous comment was ambiguous. When I suggested we set a class name on the parent list, I didn't mean use `header` for the parent list,  as it's semantically incorrect.

marta- replied:
Renamed in 90725fb.

marta- committed:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI
Refactor the display of configured survey instructions in the patient ui

marta- committed:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI
Better classname for the element listing all form entries

marta- committed:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI
Use the default instructions as placeholders

marta- committed:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI
Some refactoring:
* Rename SurveyInstConfiguration to SurveyInstrConfiguration for less ambiguity regardin the meaning of Inst(r)
* Export and reuse the survey instructions configuration path
* Use h6 instead of h5 for headings on the admin screen
* Use the same textarea hight for all configurable entries
* Add "." at the end of default instructions text

marta- committed:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI
* Refactoring: Moved the survey instructions loading logic from QuestionnaireSet.jsx to index.jsx so other components can access it
* Made some texts in the PatientIdentification component configurable

marta- committed:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI
Removed deprecated comment and unnecessary function

veronikaslc commented on proms-resources/frontend/src/main/frontend/src/proms/index.jsx:
```diff
@@ -27,10 +27,34 @@ import PatientIdentification from "./PatientIdentification.jsx";
+        setSurveyInstructions(Object.assign(DEFAULT_INSTRUCTIONS, json));
+      })
+      .catch((response) => {
+        setError(`Loading the Patient Portal Survey Instructions failed with error code ${response.status}: ${response.statusText}`);
```
Is `setError` defined?

marta- replied:
Nope, thanks for catching it. Fixed in e92f814.

marta- committed:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI
Bug fix: setError not defined in index.jsx

sashaandjic reviewed (approved):
tested all scenarios: patient login with token and without. With blanks text and without. I like the fact that `No event message` and `No surveys message` displays default value in case admin made them blank.
![image](https://user-images.githubusercontent.com/23243244/179045380-e54ef0ac-4025-4354-90f3-b8dcb1242060.png)

sdumitriu commented on proms-resources/clinical-data/src/main/resources/SLING-INF/content/Homepage/Proms/SurveyInstructions.xml:
```diff
@@ -0,0 +1,62 @@
+    </property>
+    <property>
+        <name>interpretationInstructions</name>
+        <value>There are different actions you can take now depending on how you have scoredyour symptoms. Please see below for a summary of your scores and suggested actions.</value>
```
```suggestion
        <value>There are different actions you can take now depending on how you have scored your symptoms. Please see below for a summary of your scores and suggested actions.</value>
```

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/proms/QuestionnaireSet.jsx:
```diff
@@ -363,7 +363,7 @@ function QuestionnaireSet(props) {
     return (
       <QuestionnaireSetScreen className={classes.screen}>
-        <Typography variant="h4" color="textSecondary">You do not have any pending surverys</Typography>
+        <Typography variant="h4" color="textSecondary">You do not have any pending surveys</Typography>
```
Why isn't this configured?

marta- replied:
I didn't include it it the specs as it seemed to not need to change for different usecases, unlike the other texts.

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/proms/SurveyInstrConfiguration.jsx:
```diff
@@ -0,0 +1,105 @@
+  },
+}));
+
+function SurveyInstrConfiguration() {
```
Shortening `Instructions` doesn't save that much space, it should use the full name `SurveyInstructionsConfiguratio`. For the filename as well.

marta- committed:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI

Typo fix

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

Commit:
CARDS-1836: As an administrator, I can configure the texts in the Patient Portal Survey UI
Renamed resources
